Feature/villager trades

LLytho opened the pull request:
Register modify or removing for trade offers:
`TradeRegistry.registerVillagerOfferModify(Consumer<VillagerTradeOfferContext> consumer)`
`TradeRegistry.registerVillagerOfferRemoving(Predicate<VillagerTradeOfferContext> predicate)`
`TradeRegistry.registerWanderingTraderOfferModify(Consumer<WanderingTraderOfferContext> consumer)`
`TradeRegistry.registerWanderingTraderOfferRemoving(Predicate<WanderingTraderOfferContext> predicate)`

Override the max possible offers for a villager or the wandering trader. As an example wandering trader has over 50 possible trades. Vanilla just select 5 per default.
`TradeRegistry.overrideVillagerMaxOffers(VillagerProfession profession, int level, int maxOffers)`
`TradeRegistry.overrideWanderingTraderMaxOffers(int maxOffers)`

LLytho committed:
Add modify and removing for villager trades

- Add mixin for villager trades
- Add methods to register modify and removing
- Implement base for VillagerMixin to provide additional villager data
- Basic Access & Mixin change
- Add AT and AW
- Add overriding for max offers a villager or the wanderer can have

LLytho committed:
Add rare check for wandering trader

LLytho committed:
Remove todo comment

MaxNeedsSnacks commented on common/src/main/java/me/shedaniel/architectury/registry/trade/TradeRegistry.java:
```diff
@@ -59,4 +144,38 @@ public static void registerTradeForWanderingTrader(boolean rare, VillagerTrades.
+        return WANDERING_TRADER_MAX_OFFER_OVERRIDE;
+    }
+    
+    public static boolean invokeVillagerOfferRemoving(VillagerTradeOfferContext ctx) {
```
These should not be publicly accessible to users imo, or at the very least not be included in the same class as the user-facing API. Consider moving this (as well as perhaps the maps themselves) to another class marked as internal.

LLytho replied:
With this. I would move everything into a `trading` package and just leave `MerchantOfferAccess` and the `TradeRegistry` left in the current package? And then move the invoke stuff into his own class. Something against this?

MaxNeedsSnacks replied:
I'd rather have the package be called `internal` or `impl` as we **almost definitely** don't want users to call this themselves, so highlighting this with a different package and `@ApiStatus.Internal` might help with that

MaxNeedsSnacks commented on common/src/main/java/me/shedaniel/architectury/registry/trade/TradeRegistry.java:
```diff
@@ -47,6 +59,79 @@ private static void registerVillagerTrade0(VillagerProfession profession, int le
+     * The consumer gets called when {@link net.minecraft.world.entity.npc.WanderingTrader} generates their offer list.
+     * @param consumer The consumer to handle modification for the given offer context.
+     */
+    public static void registerWanderingTraderOfferModify(Consumer<WanderingTraderOfferContext> consumer) {
```
```suggestion
    public static void modifyWanderingTraderOffers(Consumer<WanderingTraderOfferContext> callback) {
```

MaxNeedsSnacks commented on common/src/main/java/me/shedaniel/architectury/registry/trade/TradeRegistry.java:
```diff
@@ -47,6 +59,79 @@ private static void registerVillagerTrade0(VillagerProfession profession, int le
+     * The predicate gets called when {@link net.minecraft.world.entity.npc.Villager} generates their offer list.
+     * @param predicate The predicate to test if an offer should be removed. Returning true means the offer will be removed.
+     */
+    public static void registerVillagerOfferRemoving(Predicate<VillagerTradeOfferContext> predicate) {
```
```suggestion
    public static void removeVillagerOffers(Predicate<VillagerTradeOfferContext> filter) {
```

MaxNeedsSnacks commented on common/src/main/java/me/shedaniel/architectury/registry/trade/TradeRegistry.java:
```diff
@@ -47,6 +59,79 @@ private static void registerVillagerTrade0(VillagerProfession profession, int le
+     * The consumer gets called when {@link net.minecraft.world.entity.npc.Villager} generates their offer list.
+     * @param consumer The consumer to handle modification for the given offer context.
+     */
+    public static void registerVillagerOfferModify(Consumer<VillagerTradeOfferContext> consumer) {
```
```suggestion
    public static void modifyVillagerOffers(Consumer<VillagerTradeOfferContext> callback) {
```

MaxNeedsSnacks commented on common/src/main/java/me/shedaniel/architectury/registry/trade/TradeRegistry.java:
```diff
@@ -47,6 +59,79 @@ private static void registerVillagerTrade0(VillagerProfession profession, int le
+     * @param level      The level of the villager. Vanilla range is 1 to 5, however mods may extend that upper limit further.
+     * @param maxOffers  Max possible offers a villager can have.
+     */
+    public static void overrideVillagerMaxOffers(VillagerProfession profession, int level, int maxOffers) {
```
```suggestion
    public static void setVillagerMaxOffers(VillagerProfession profession, int level, int maxOffers) {
```

MaxNeedsSnacks commented on common/src/main/java/me/shedaniel/architectury/registry/trade/TradeRegistry.java:
```diff
@@ -59,4 +144,38 @@ public static void registerTradeForWanderingTrader(boolean rare, VillagerTrades.
+     * Override the max possible offers the wandering trader can have. This does not affect the rare trade.
+     * @param maxOffers  Max possible offers a villager can have.
+     */
+    public static void overrideWanderingTraderMaxOffers(int maxOffers) {
```
```suggestion
    public static void setWanderingTraderMaxOffers(int maxOffers) {
```

MaxNeedsSnacks commented on common/src/main/java/me/shedaniel/architectury/registry/trade/TradeRegistry.java:
```diff
@@ -47,6 +59,79 @@ private static void registerVillagerTrade0(VillagerProfession profession, int le
+     * The predicate gets called when {@link net.minecraft.world.entity.npc.WanderingTrader} generates their offer list.
+     * @param predicate The predicate to test if an offer should be removed. Returning true means the offer will be removed.
+     */
+    public static void registerWanderingTraderOfferRemoving(Predicate<WanderingTraderOfferContext> predicate) {
```
```suggestion
    public static void removeWanderingTraderOffers(Predicate<WanderingTraderOfferContext> filter) {
```

MaxNeedsSnacks reviewed (changes requested):
Only managed to have a cursory glance at these as I'm a bit strapped for time, so this isn't a full review yet, but I think these things may need addressing

LLytho committed:
rename some methods

MaxNeedsSnacks commented:
Sorry for the long wait, I'll be checking this out soon ^^
(That said @shedaniel also seems to have taken interest in it seeing as he self-assigned this, so I might also just wait for his review here)

shedaniel commented on common/src/main/java/me/shedaniel/architectury/registry/trade/TradeRegistry.java:
```diff
@@ -23,7 +23,19 @@
+    private static final List<Predicate<WanderingTraderOfferContext>> WANDERING_TRADER_REMOVE_HANDLERS = new ArrayList<>();
+    
+    private static final Map<VillagerProfession, Map<Integer, Integer>> VILLAGER_MAX_OFFER_OVERRIDES = new HashMap<>();
+    private static Integer WANDERING_TRADER_MAX_OFFER_OVERRIDE = null;
```
wrong case

LLytho replied:
just for `WANDERING_TRADER_MAX_OFFER_OVERRIDE` here right?

shedaniel replied:
yeah because it is not static final

shedaniel commented on common/src/main/java/me/shedaniel/architectury/registry/trade/TradeRegistry.java:
```diff
@@ -59,4 +144,38 @@ public static void registerTradeForWanderingTrader(boolean rare, VillagerTrades.
+    /**
+     * @return Max offers for the wandering trader. Returning null means no override exists
+     */
+    public static Integer getWanderingTraderMaxOffers() {
```
getWanderingTraderMaxOffersOverride or something along the lines of that but shorter

also @Nullable

LLytho replied:
Without just shorten `WanderingTrader` into `WT` what would be a shorter version for it? I don't think in general longer method names are bad.

 added @nullable for `getVillagerMaxOffers` too

shedaniel replied:
tbh considering this again maybe it is best to not expose this method in the "api", using WT is kinda meh and I don't feel like that's where we want to simplify at all, would 100% want the full WanderingTrader instead

shedaniel commented on common/src/main/java/me/shedaniel/architectury/mixin/WanderingTraderMixin.java:
```diff
@@ -0,0 +1,94 @@
+    
+    @Override
+    public int overrideMaxOffersIfRegistered(int currentMaxOffers) {
+        Integer newMaxOffers = TradeRegistry.getWanderingTraderMaxOffers();
```
could use MoreObjects.firstNonNull

shedaniel commented on common/src/main/java/me/shedaniel/architectury/mixin/AbstractVillagerMixin.java:
```diff
@@ -0,0 +1,87 @@
+        return handledOffer;
+    }
+    
+    public MerchantOffer handleOfferImpl(MerchantOffer offer) {
```
this needs to have architectury$ prefixed

shedaniel commented on common/src/main/java/me/shedaniel/architectury/registry/trade/MerchantOfferAccess.java:
```diff
@@ -0,0 +1,79 @@
+        this.offer = offer;
+    }
+    
+    public ItemStack getPrimary() {
```
these should be getPrimaryCost or exactly same as the mojmap ones

LLytho replied:
```java
public class MerchantOffer {
   public ItemStack baseCostA;
   public ItemStack costB;
   public ItemStack result;
   // ...
}
```

So better `getCostA` and `getCostB` or maybe `getPrimaryPrice` like in `SimpleTrade`?

shedaniel replied:
I would use cost instead of price here, but basically yes

shedaniel commented on common/src/main/java/me/shedaniel/architectury/registry/trade/MerchantOfferAccess.java:
```diff
@@ -0,0 +1,79 @@
+import net.minecraft.world.item.trading.MerchantOffer;
+
+public class MerchantOfferAccess {
+    private final MerchantOffer offer;
```
might want a getter for this field

shedaniel reviewed (commented):
Pretty good, some minor issues, will test this locally later

LLytho committed:
Solve reviews for #122

Move non api stuff into TradeRegistryData
Rename fields in MerchantOfferAccess
Move trade stuff into internal package
Mark internal trade classes as ApiStatus.Internal

LLytho committed:
Merge branch '1.16' into feature/villager-trades

MaxNeedsSnacks committed:
Minor refactors (discussed on Discord)

LLytho committed:
Add doc for AbstractVillagerMixin

LLytho committed:
Reformat code

MaxNeedsSnacks committed:
Update gradle.properties
